### ISSUE TITLE
multi-gpu + different image sizes ?

### ISSUE THREAD

**Issue by RSly** [NVIDIA/caffe#335]
Hi,

I am trying to train a FCN with multi-gpu training (both in nv-caffe 0.15 and 0.16) on pascalvoc dataset which has different image sizes. I use batch_size =1 for each GPU (since we cant load multiple varying size images in one batch), and 2-4 GPUs.

in each case I get memory errors, which is not really due to memory being full (image dimensions are <500) also all works fine in 1-gpu training with no memory errors. this looks like a bug to me that doesnt consider varying-size images in multi-gpu case

so I have a couple of questions:
* Does nv-caffe support multi-gpu training for varying image sizes for FCNs? example one gpu is working on 500x300 image and the other on 400x200 ? 
* is the above behavior a bug?

p.s. I know we can resize/pad the images to a fixed size, but thanks to fcn this is not a requirement

**Comment by drnikolaev** [comment]
@RSly could you copy your net.prototxt here? We just need to check something before we dig deeper to this.

**Comment by RSly** [comment]
hi @drnikolaev 
* I pulled the last day changes and for some magical reason it all works now with caffe 0.16, multi-gpu, varied size images

* I also confirm that none of these experiments could run with caffe 0.15, so it looks like a big step forward.

* I still notice  that the attached 81 class proto crashes for memory reasons in digits, which shouldn't be the case with a 12GB gpu, since the memory consumption should be around 4GB in average with image sizes <500 (actually it doesnt crash using a direct python datalayer instead of digits LMDB)
[fcn_alexnet_81.txt](https://github.com/NVIDIA/caffe/files/1003619/fcn_alexnet_81.txt)


F0516 09:29:08.443459 31196 gpu_memory.hpp:38] Failed to allocate 3514368 bytes on device 1. Total memory: 12799574016, Free: 589824, dev_info[1]: total=12799574016 free=589824
*** Check failure stack trace: ***
@     0x7ff515665daa  (unknown)
@     0x7ff515665ce4  (unknown)
@     0x7ff5156656e6  (unknown)
@     0x7ff515668687  (unknown)
@     0x7ff515e1a3d0  caffe::GPUMemory::allocate<>()
@     0x7ff515e1913c  caffe::SyncedMemory::to_gpu()
@     0x7ff515e1a03d  caffe::SyncedMemory::mutable_gpu_data()
@     0x7ff515e0400c  caffe::Blob::PtrProxy<>::operator float()
@     0x7ff516253c38  caffe::CuDNNConvolutionLayer<>::Backward_gpu()
@     0x7ff515e860e5  caffe::Layer<>::Backward()
@     0x7ff515e428c1  caffe::Net::BackwardFromToAu()
@     0x7ff515e42bd5  caffe::Net::Backward()
@     0x7ff515e42e51  caffe::Net::ForwardBackward()
@     0x7ff515ea40f3  caffe::Solver::Step()
@     0x7ff515ea4e00  caffe::Solver::Solve()
@           0x40d96f  train()
@           0x40aea4  main
@     0x7ff513e5ef45  (unknown)
@           0x40b77b  (unknown)
@              (nil)  (unknown)

**Comment by RSly** [comment]
Hi @drnikolaev 
Here is an update on this: the behavior is random
=> I could run the same experiment with the above net.proto, it sometimes kicks off and trains normally with a steady GPU RAM usage ~ 4gb for over 5000 iterations, on other times after some iterations the usage suddenly jumps to 12Gb and then it crashes as the following

I0516 11:27:37.752105  3754 sgd_solver.cpp:108] Iteration 4, lr = 0.000999593
I0516 11:27:38.121021  3754 solver.cpp:300] Iteration 5 (2.71087 iter/s, 0.368885s/1 iter), loss = 0.538171
I0516 11:27:38.121062  3754 solver.cpp:322]     Train net output #0: accuracy-81c = 0.888239
I0516 11:27:38.121073  3754 solver.cpp:322]     Train net output #1: smloss-81c = 0.538171 (* 1 = 0.538171 loss)
I0516 11:27:38.121078  3754 sgd_solver.cpp:108] Iteration 5, lr = 0.000999491
F0516 11:27:38.596923  3755 gpu_memory.hpp:38] Failed to allocate 121352256 bytes on device 2. Total memory: 12799574016, Free: 59047936, dev_info[2]: total=12799574016 free=59047936
*** Check failure stack trace: ***
    @     0x7f5190f9edaa  (unknown)
    @     0x7f5190f9ece4  (unknown)
    @     0x7f5190f9e6e6  (unknown)
    @     0x7f5190fa1687  (unknown)
    @     0x7f51917533d0  caffe::GPUMemory::allocate<>()
    @     0x7f519175213c  caffe::SyncedMemory::to_gpu()
    @     0x7f519175303d  caffe::SyncedMemory::mutable_gpu_data()
    @     0x7f519173d00c  caffe::Blob::PtrProxy<>::operator float()
    @     0x7f5191aaf60d  caffe::SplitLayer<>::Backward_gpu()
    @     0x7f51917bf0e5  caffe::Layer<>::Backward()
    @     0x7f519177b8c1  caffe::Net::BackwardFromToAu()
    @     0x7f519177bbd5  caffe::Net::Backward()
    @     0x7f519177be51  caffe::Net::ForwardBackward()
    @     0x7f51917dd0f3  caffe::Solver::Step()
    @     0x7f51917dde00  caffe::Solver::Solve()
    @     0x7f51917341eb  caffe::P2PSync::InternalThreadEntry()
    @     0x7f519173fa82  caffe::InternalThread::entry()
    @     0x7f51917406f4  boost::detail::thread_data<>::run()
    @     0x7f518f56ba4a  (unknown)
    @     0x7f518f34a184  start_thread
    @     0x7f518f87037d  (unknown)
    @              (nil)  (unknown)

**Comment by RSly** [comment]
I attached the gpu ram usages for 4 different runs of the exact same experiment with the above alexnetfcn and 21 classes, you can see the usage is random
=> run 2 seems to have the 'good' behavior
=> run 4 results in immediate crash ~ 12gb usage on gpu1
=> other runs also use heavily  the ram for no reason
![gpu-usage](https://cloud.githubusercontent.com/assets/3428292/26100494/580f3170-3a2e-11e7-95c6-b4b3ef7fb1f0.png)

**Comment by drnikolaev** [comment]
Hi @RSly , good catch, thank you. And perfect timing - I was optimizing memory balancer.
Two challenges addressed:

1. FindEx mechanism computes maximum memory amount which _might be_ needed but fastest algorithm needs much less. The bug was just about missing clean up...
2. It's not a good idea to take 9GB out of 12. Fixed.

Merge candidate is here (it's being tested): https://github.com/drnikolaev/caffe/tree/caffe-0.16
You are very welcome to give it a try.

**Comment by RSly** [comment]
hi @drnikolaev 
thanks for the update. I ran the tests from my side!
=> it looks very good!

the only issues I saw are as following:
* on a test-deck where the normal ram usage is < 4G, in the initialization phase and for a short moment, the usage on one gpu jumps to 7GB and very fast goes back to <4G, this can cause a crash and looks like a clean up bug

* also check this out, there is something going on with the reallocating workspace. The same run work perfectely with 1 gpu, but crashes as following with 4 gpus:

I0519 10:41:06.080021  3719 cudnn_conv_layer.cpp:834] [3] Conv Algos (F,BD,BF): 'score_fr-81c' with space 4.58G/1 2 1 0  (limit 4.57G, req 0.03G)
I0519 10:41:06.088117  3716 cudnn_conv_layer.cpp:834] [0] Conv Algos (F,BD,BF): 'score_fr-81c' with space 4.58G/1 2 1 0  (limit 4.25G, req 0.03G)
I0519 10:41:06.488883  3716 solver.cpp:300] Iteration 2 (0.605592 iter/s, 1.65128s/5000 iter), loss = 4.16255
I0519 10:41:06.488940  3716 solver.cpp:322]     Train net output #0: accuracy-81c = 0.613027
I0519 10:41:06.488950  3716 solver.cpp:322]     Train net output #1: smloss-81c = 4.16255 (* 1 = 4.16255 loss)
I0519 10:41:06.489204  3716 cudnn_conv_layer.cpp:282] [0] Layer 'conv1' reallocating workspace: 4.58G -> 0.05G
I0519 10:41:06.489437  3719 cudnn_conv_layer.cpp:282] [3] Layer 'conv1' reallocating workspace: 4.58G -> 0.05G
I0519 10:41:06.492197  3718 cudnn_conv_layer.cpp:282] [2] Layer 'conv1' reallocating workspace: 0.02G -> 0.01G
F0519 10:43:37.118813  3771 syncedmem.cpp:185] Check failed: error == cudaSuccess (11 vs. 0)  invalid argument
*** Check failure stack trace: ***
@     0x7fdc47eaedaa  (unknown)
@     0x7fdc47eaece4  (unknown)
@     0x7fdc47eae6e6  (unknown)
@     0x7fdc47eb1687  (unknown)
@     0x7fdc48667689  caffe::SyncedMemory::async_gpu_push()
@     0x7fdc487fd93a  caffe::BasePrefetchingDataLayer<>::InternalThreadEntryN()
@     0x7fdc48653e61  caffe::InternalThread::entry()
@     0x7fdc48654bc4  boost::detail::thread_data<>::run()
@     0x7fdc4647ba4a  (unknown)
@     0x7fdc4625a184  start_thread
@     0x7fdc4678037d  (unknown)
@              (nil)  (unknown)

**Comment by RSly** [comment]
one question though, can we use batch_size>1 when the image sizes in the batch are not the same?
currently, it is not supported and we end up with a crash as shown below.
=>maybe internally padding the images to the largest size can be an option...
=> I actually dont know any good solution for relaxing the fixed size requirement in the batch.

any suggestions out there?

F0519 10:31:12.084918  2532 syncedmem.cpp:185] Check failed: error == cudaSuccess (11 vs. 0)  invalid argument
*** Check failure stack trace: ***
@     0x7f4d30e00daa  (unknown)
@     0x7f4d30e00ce4  (unknown)
@     0x7f4d30e006e6  (unknown)
@     0x7f4d30e03687  (unknown)
@     0x7f4d315b9689  caffe::SyncedMemory::async_gpu_push()
@     0x7f4d3174f93a  caffe::BasePrefetchingDataLayer<>::InternalThreadEntryN()
@     0x7f4d315a5e61  caffe::InternalThread::entry()
@     0x7f4d315a6bc4  boost::detail::thread_data<>::run()
@     0x7f4d2f3cda4a  (unknown)
@     0x7f4d2f1ac184  start_thread
@     0x7f4d2f6d237d  (unknown)
@              (nil)  (unknown)

**Comment by drnikolaev** [comment]
@RSly thanks for checking. As of 7GB. Let me briefly clarify. CuDNN v6 comes packed with good convolution algorithms, and NVCaffe tries to choose the fastest one. Before trying, it checks how much memory would every candidate need to do the job and allocates one workspace for them all (in your case it's 7GB). Then it starts short races and choose the winner. If winner needs less than 7GB NVCaffe shrinks the workspace to the size the winner wants. I'm thinking about some user setting limiting memory consumption. For example, you want NVCaffe to use not more than 5GB. Would it be useful?
As of the crash - I can't reproduce it yet but I'll keep trying
As of different images - yes, you do need to set batch size to 1 but in this case multi-GPU mode wouldn't make much sense (I have to double check it though). Does single GPU mode work fine on your side?

**Comment by RSly** [comment]
hi @drnikolaev 

- regarding the ram usage, I summarized my observation in https://github.com/NVIDIA/caffe/issues/341

- for the crash, yes single gpu works fine, it happens in multi-gpu and only in combination with digits, and does not happen with python data layer

- regarding the batsh_size=1 and multi GPU, it all works fine currently! Also I think it is a very useful feature in some cases: eg. one has large images sizes that already fills one gpu ram, so basically x4 gpus is a great natural way to go, otherwise we need to crop and stitch and re-balance all the images, etc.
